### PR TITLE
Handling edge cases on `heading` and `list_item` nodes

### PR DESCRIPTION
**Pull request by Edo-San**
Hey @edvinasjurele 👋 
First of all, I want to thank you for this amazing package! The effort you put into this repo is outstanding!

I work at Storyblok in the Website team, and we were on the verge of needing to develop something like this. Since this work from yours is truly astonishing IMHO, we've been trying it out ourselves and it's working like a charm.

However, I found a couple of bugs that I've been trying to assess with this PR.

### The issue
The `Heading` content is not marked as optional, while, in fact, it is. Set aside the fact that an empty heading doesn't mean anything, there may be situations where an editor could inadvertently add a heading node to a Richtext field (which is as easy as clicking on the heading button) and then proceed to forget about it 😅 

The same thing goes for `Paragraph`s, but while handling the unhappy path is a thing in the case of plain paragraphs, it is not true for paragraphs rendered inside `List Item`s. In those cases, if a paragraph doesn't have any content, `node.content` is `undefined` and therefore an exception is thrown.

### The proposed solution
For `content`less headings: I thought it could be reasonable to render a `<br />` instead of the actual heading, much like the `paragraph` implementation, but maybe it could be better not to render anything at all?
For `p`s inside list items: it doesn't feel "right" to me to render a `<br />` in this case, as it could very well lead to quirks that may be avoided by not rendering anything 🤔
The only thing I want to point out here is that returning an empty string as content would potentially mean that empty strings could be passed as a parameter to the `renderNode` function, which is meant to only accept `ComponentNode`

I'm open to suggestions.
Thanks again for your amazing work! ❤️

### CODE REVIEW

**Comment by edvinasjurele** [comment]
Hi @Edo-San, thanks for your MR and the kind words! I am happy to help and give back to the community 🙏 

I will review the code and propose the improvements (if any).

--- 
To start with, I see DCO is failing, it means you need to commit stuff with `--signoff` (or `-s`) flag (see [more info](https://github.com/apps/dco)). Easy as that:
```
git commit -s -m 'This is my commit message'
```
You may follow the [details](https://github.com/NordSecurity/storyblok-rich-text-astro-renderer/pull/65/checks?check_run_id=16514876405) to resolve that.

--- 
Secondly, we might need to do the business decisions:
1. **What to render instead of empty Heading?** Totally agree with you to render the `<br />` (same as for paragraph, btw, those might be used for making "empty lines" when writing a blog article), thus we might treat empty headings the same way, leaving this empty heading noticable in both app view and Storyblok rich text view, thus it resembles nicely and leaves possibility for content editor to spot it easier and remove redundant empty headings:

image

(in image, the first gap is empty heading, the second gap is empty paragraph)

2. **What to render instead of empty list item?** Yes, we may render out simple empty string `""`, respecting the content editor choice of it being empty. And yes, in all cases we skip rendering that paragraph in list items.

image

Although, just FYI, there is one case where we should still respect hard breaks in list item and that seems logic to me and works like this already (nothing to fix):

image

 .
3. **What to do with `renderNode` function potentially getting empty string?** We have to make sure any node or array of nodes contains another node (maintain node-in-node structure), thus the `[""]` must become `[{ content: "" }]` and the resolvedContent will land as simple "". I hope I got this right, or let me know if I miss anything on this aspect.

File: lib/src/utils/resolveRichTextToNodes.spec.ts
Comment (edvinasjurele) [inline lib/src/utils/resolveRichTextToNodes.spec.ts:240]: it seems we have here `describe in describe` situation. it is not quite necessarily bad, but inconsistent at the moment.

The whole file is done in this structure:
```
describe('resolveNode', () => {
  it('heading', () => {
    expect(...).toStrictEqual({ ... });
    expect(...).toStrictEqual({ ... });
    expect(...).toStrictEqual({ ... });
  });
  it('paragraph', () => { ... });
  it('hard_break', () => { ... });
  ...
});

describe('resolveMark', () => {
  it('link', () => { ... });
  it('bold', () => { ... });
  ...
});
```

so following that you might simply go with:

```
// empty content
expect(resolveNode(emptyNode)).toStrictEqual({
   component: "br",
 });
```
or

```
// empty content
expect(resolveNode({
      type: "heading",
      attrs: {
        level: 2,
      },
    })).toStrictEqual({
   component: "br",
 });
```
to be extra concise.

File: lib/src/utils/resolveRichTextToNodes.spec.ts
Comment (edvinasjurele) [inline lib/src/utils/resolveRichTextToNodes.spec.ts:427]: You should maintain node-in-node here, thus `content: [""]` => `content: [{ content: "" }]` so we have valid nodes nested. See `renderNode` function, as if it recognizes the array, it goes though all the nodes inside it: https://github.com/NordSecurity/storyblok-rich-text-astro-renderer/blob/main/lib/RichTextRenderer.astro#L39

Btw, we should have this case too, so maybe we should add an extra case for it (not sure if this gives extra value):
image

File: lib/src/utils/resolveRichTextToNodes.ts
Comment (edvinasjurele) [inline lib/src/utils/resolveRichTextToNodes.ts:265]: Having in mind the other comment about `[""]` case, you may go with the following instead:

```
return {
  content: node.content?.map((node) => resolveNode(node, options)) || "",
};
```

**Review by edvinasjurele — CHANGES_REQUESTED** [review]
Nice work on catching these edge cases 💪

**Comment by edvinasjurele** [comment]
> I saw the DCO error when I opened the PR 🤦 Will re-commit ASAP.
> 
> It all looks and sounds good on my side! As for the last bit, I'd just amend [this commit](https://github.com/NordSecurity/storyblok-rich-text-astro-renderer/pull/65/commits/fa6f439ae18b1f2e4acdf4a22a3ef39b8bc79850), WDYT? I implemented that bit as you described in this commit, and it's probably better to stick with it as it wouldn't require messing with typings as you said.

Sure, formulate the commits in any way you want, just maintain the separation of concerns (if possible). As the each commits is indeed a changelog entry in the releases notes https://github.com/NordSecurity/storyblok-rich-text-astro-renderer/releases

File: lib/src/utils/resolveRichTextToNodes.spec.ts
Comment (Edo-San) [reply lib/src/utils/resolveRichTextToNodes.spec.ts:239]: Yup, totally! It could be an improvement in terms of failed testing readability, but that's definitely out of the scope of this PR.

**Comment by Edo-San** [comment]
Hey @edvinasjurele thanks for your feedback, much appreciated! 🙌 
As for the commits for the release, did you configure GitHub to squash-commit when merging PRs? It could be beneficial in terms of hiding bunches of rather useless commits in favor of changelog/releaselog readability (it should automatically create the squash commit message based on the title of the PR, if I'm not mistaken).

Otherwise, I'll squash everything myself :D

**Review by edvinasjurele — APPROVED** [review]

**Comment by edvinasjurele** [comment]
> Hey @edvinasjurele thanks for your feedback, much appreciated! 🙌 As for the commits for the release, did you configure GitHub to squash-commit when merging PRs? It could be beneficial in terms of hiding bunches of rather useless commits in favor of changelog/releaselog readability (it should automatically create the squash commit message based on the title of the PR, if I'm not mistaken).
> 
> Otherwise, I'll squash everything myself :D

No we haven't and here is the reason, and I go with `Rebase and merge`. As mentioned in [contributing](https://github.com/NordSecurity/storyblok-rich-text-astro-renderer/blob/main/README.md#contributing), we do follow semantic-release practices, and respect the contributor commits as is, leaving a space to add various changes in a single PR, f.e.:
```
fix: handle undefined content provided to heading node
docs: update README
chore: bump storyblok/astro package
```

But at the same time maintinaining proper changelog. From those commits the tool generates the changelog and picks the next tag version automatically.
To be honest, I haven't yet finalised semantic-release configuration, thus I am doing those releases manually in this project for now (I will have to find some time to finish CI stuff as I am not used to Github one 🙂 ).

Two ways from here:
1. Please squash your commits into one  🙏 (it probably makes sense in this PR to have a single one as all of those are about the same problem to fix).
2. It could be those first 4 commits (those are totally fine) but the last one is a little off. Might have been squashed to one of the existing ones and we are ready to go.

image

Your call!

**Comment by edvinasjurele** [comment]
I have released the new package version https://www.npmjs.com/package/storyblok-rich-text-astro-renderer 🎉 
Thank you again for the contribution!

Btw, let's meet on Discord. My user is eddyJ#6064

**Pull request merged**
